components/feature/webchannel/src/main/java/mozilla/components/feature/webchannel/WebChannelViewFeature.kt

```diff
@@ -0,0 +1,201 @@
+import mozilla.components.concept.engine.webextension.WebExtension
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.base.log.logger.Logger
+import org.json.JSONArray
+import org.json.JSONObject
+import java.util.WeakHashMap
+
+/**
+ * Feature implementation that provides Firefox Accounts WebChannel support.
+ * This feature is implemented as a web extension and
+ * needs to be installed prior to use (see [WebChannelViewFeature.install]).
+ *
+ * @property context a reference to the context.
+ * @property engine a reference to the application's browser engine.
+ * @property sessionManager a reference to the application's [SessionManager].
+ */
+@Suppress("TooManyFunctions")
+class WebChannelViewFeature(
+    private val context: Context,
+    private val engine: Engine,
+    private val sessionManager: SessionManager
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature {
+
+    override fun start() {
+        observeSelected()
+
+        registerContentMessageHandler(activeSession)
+
+        if (WebChannelViewFeature.installedWebExt == null) {
+            WebChannelViewFeature.install(engine)
+        }
+    }
+
+    override fun stop() {
+        super.stop()
+    }
+
+    override fun onSessionSelected(session: Session) {
+        super.onSessionSelected(session)
+    }
+
+    override fun onSessionAdded(session: Session) {
+        registerContentMessageHandler(session)
+    }
+
+    override fun onSessionRemoved(session: Session) {
+        ports.remove(sessionManager.getEngineSession(session))
+    }
+
+    private fun sendStatusResponse(messageId: String) {
+        val statusData = JSONObject()
+        statusData.put("signedInUser", "testuser@testuser.com")
+        statusData.put("sessionToken", "ya")
+        statusData.put("uid", "uid")
+        statusData.put("verified", true)
+        statusData.put("capabilities", JSONObject().put("engines", JSONArray().put("creditcards")))
+
+        val statusMessage = JSONObject()
+        statusMessage.put("messageId", messageId)
+        statusMessage.put("command", COMMAND_STATUS)
+        statusMessage.put("data", statusData)
+
+        val status = JSONObject()
+        status.put("id", CHANNEL_ID)
+        status.put("message", statusMessage)
+
+        sendContentMessage(status)
+    }
+
+    private fun sendLinkResponse(messageId: String) {
+        val statusData = JSONObject()
+        statusData.put("ok", true)
+
+        val statusMessage = JSONObject()
+        statusMessage.put("messageId", messageId)
+        statusMessage.put("command", COMMAND_CAN_LINK_ACCOUNT)
+        statusMessage.put("data", statusData)
+
+        val status = JSONObject()
+        status.put("id", CHANNEL_ID)
+        status.put("message", statusMessage)
+
+        sendContentMessage(status)
+    }
+
+    private fun recieveLogin(messageObj: JSONObject) {
+        logger.info(messageObj.toString())
```

Comment:
the result of this here is something like 

> {"messageId":null,"data":{"verified":false,"keyFetchToken":"7f6ef1a55cc2e629c5b587c5c1fe2a0e8e457ce594e767072e5fbe356b47bba5","sessionToken":"02c354774aa81accae09ada17b21816436dc5461250a60800d078d843c5155e2","unwrapBKey":"cbfabce2b955c06b8057f66b23073efd7fcf3790db5c2967e2291ed1740aafa8","uid":"e82283bfcc4d4e65860e4dc6ea82d378","email":"v@v.com","customizeSync":false,"verifiedCanLinkAccount":true},"command":"fxaccounts:login"}

Reply:
@Amejia481 @grigoryk new event is now called `oauth_login` and has this right now:

```


{"messageId":null,"data":{"action":"signup","redirect":"http:\/\/127.0.0.1:3030\/oauth\/success\/a2270f727f45f648?code=0fdbc2e14ffa3b1d933c3d802aa4753eb6d781c38c0bd67981b6a69fa0623619&state=vHao1p6OizzwReCkQMSpZA","declinedSyncEngines":["history"],"code":"0fdbc2e14ffa3b1d933c3d802aa4753eb6d781c38c0bd67981b6a69fa0623619","offeredSyncEngines":["bookmarks","history","passwords","addons","tabs","prefs"],"state":"vHao1p6OizzwReCkQMSpZA","customizeSync":true},"command":"fxaccounts:oauth_login"}
```